src/Migration.php

```diff
@@ -0,0 +1,583 @@
+ *
+ * Note that some DB queries in some DBMS cannot be put into a transaction. For some examples, please refer to
+ * [implicit commit](http://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html). If this is the case, you should still
+ * implement `safeUp()` and `safeDown()`, instead.
+ *
+ * Migration provides a set of convenient methods for manipulating database data and schema.
+ *
+ * For example, the {@see insert()} method can be used to easily insert a row of data into a database table; the
+ * {@see createTable()} method can be used to create a database table.
+ *
+ * Compared with the same methods in {@see Command}, these methods will display extra information showing the method
+ * parameters and execution time, which may be useful when applying migrations.
+ *
+ * For more details and usage information on Migration, see the [guide article on Migration](guide:db-migrations).
+ */
+abstract class Migration implements MigrationInterface
+{
+    use SchemaBuilderTrait;
+
+    private int $maxSqlOutputLength = 0;
+    private bool $compact = false;
+    protected Connection $db;
+
+    public function __construct(Connection $db)
+    {
+        $this->db = $db;
+        $this->db->getSchema()->refresh();
+        $this->db->setEnableSlaves(false);
+    }
+
+    public function getDb(): Connection
+    {
+        return $this->db;
+    }
+
+    /**
+     * This method contains the logic to be executed when applying this migration.
+     *
+     * This method differs from {@see up()} in that the DB logic implemented here will be enclosed within a DB
+     * transaction.
+     * Child classes may implement this method instead of {@see up()} if the DB logic needs to be within a transaction.
+     *
+     * Note: Not all DBMS support transactions. And some DB queries cannot be put into a transaction. For some examples,
+     * please refer to [implicit commit](http://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html).
+     *
+     * @return bool return a false value to indicate the migration fails and should not proceed further. All other
+     * return values mean the migration succeeds.
+     */
+    public function safeUp()
+    {
+    }
+
+    /**
+     * This method contains the logic to be executed when removing this migration.
+     *
+     * This method differs from {@see down()} in that the DB logic implemented here will be enclosed within a DB
+     * transaction.
+     * Child classes may implement this method instead of {@see down()} if the DB logic needs to be within a
+     * transaction.
+     *
+     * Note: Not all DBMS support transactions. And some DB queries cannot be put into a transaction. For some examples,
+     * please refer to [implicit commit](http://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html).
+     *
+     * @return bool return a false value to indicate the migration fails and should not proceed further. All other
```

Comment:
Is there any reason to return non-boolean value here?

Reply:
Fix phpdoc.